pkg/ccl/backupccl/backup_metadata.go

```diff
@@ -0,0 +1,1135 @@
+
+		case bytes.HasPrefix(k.Key, []byte(sstFilesPrefix)):
+			spanStart, path, err := decodeUnsafeFileSSTKey(k.Key)
+			if err != nil {
+				return err
+			}
+			f, err := pbBytesToJSON(iter.UnsafeValue(), &BackupManifest_File{})
+			if err != nil {
+				return err
+			}
+			if err := out(k.String(), fmt.Sprintf("file %s (%s)", path, spanStart.String()), f); err != nil {
+				return err
+			}
+		case bytes.HasPrefix(k.Key, []byte(sstNamesPrefix)):
+			db, sc, name, err := decodeUnsafeNameSSTKey(k.Key)
+			if err != nil {
+				return err
+			}
+			var id uint64
+			if v := iter.UnsafeValue(); len(v) > 0 {
+				_, id, err = encoding.DecodeUvarintAscending(v)
+				if err != nil {
+					return err
+				}
+			}
+			mapping := fmt.Sprintf("name db %d / schema %d / %q @ %v -> %d", db, sc, name, k.Timestamp, id)
+			if err := out(k.String(), mapping, nil); err != nil {
+				return err
+			}
+
+		case bytes.HasPrefix(k.Key, []byte(sstSpansPrefix)):
+			span, err := decodeSpanSSTKey(k.Key)
+			if err != nil {
+				return err
+			}
+			if err := out(k.String(), fmt.Sprintf("span %s @ %v", span, k.Timestamp), nil); err != nil {
+				return err
+			}
+
+		case bytes.HasPrefix(k.Key, []byte(sstStatsPrefix)):
+			tblID, statID, err := decodeStatSSTKey(k.Key)
+			if err != nil {
+				return err
+			}
+			s, err := pbBytesToJSON(iter.UnsafeValue(), &stats.TableStatisticProto{})
+			if err != nil {
+				return err
+			}
+			if err := out(k.String(), fmt.Sprintf("stats tbl %d, id %d", tblID, statID), s); err != nil {
+				return err
+			}
+
+		case bytes.HasPrefix(k.Key, []byte(sstTenantsPrefix)):
+			id, err := decodeTenantSSTKey(k.Key)
+			if err != nil {
+				return err
+			}
+			i, err := pbBytesToJSON(iter.UnsafeValue(), &descpb.TenantInfo{})
+			if err != nil {
+				return err
+			}
+			if err := out(k.String(), fmt.Sprintf("tenant %d", id), i); err != nil {
+				return err
+			}
+
+		default:
+			if err := out(k.String(), "unknown", json.FromString(fmt.Sprintf("%q", iter.UnsafeValue()))); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// BackupMetadata holds all of the data in BackupManifest except a few repeated
+// fields such as descriptors or spans. BackupMetadata provides iterator methods
+// so that the excluded fields can be accessed in a streaming manner.
+type BackupMetadata struct {
+	BackupManifest
+	store    cloud.ExternalStorage
+	enc      *jobspb.BackupEncryptionOptions
+	filename string
+}
+
+func newBackupMetadata(
+	ctx context.Context,
+	exportStore cloud.ExternalStorage,
+	sstFileName string,
+	encryption *jobspb.BackupEncryptionOptions,
+) (*BackupMetadata, error) {
+	var encOpts *roachpb.FileEncryptionOptions
+	if encryption != nil {
+		key, err := getEncryptionKey(ctx, encryption, exportStore.Settings(), exportStore.ExternalIOConf())
+		if err != nil {
+			return nil, err
+		}
+		encOpts = &roachpb.FileEncryptionOptions{Key: key}
+	}
+
+	iter, err := storageccl.ExternalSSTReader(ctx, exportStore, sstFileName, encOpts)
```

Comment:
I think this is missing a Close(), either deferred right here, or potentially stored in BackupMetadata and closed in a separate user-called Close if we pursue my point below about just having the typed iters seek and read from this `iter` directly.

Reply:
Done, added close